src/plugins/data/common/search/search_source/search_source.ts

```diff
@@ -918,19 +860,188 @@ export class SearchSource {
+      setWith(body, '_source.includes', remainingFields, (nsValue) => {
+        return isObject(nsValue) ? {} : nsValue;
+      });
     }
 
-    const omitByIsNil = (object: Record<string, any>) => omitBy(object, isNil);
+    const builtQuery = this.getBuiltEsQuery({
+      index,
+      query: searchRequest.query,
+      filters: searchRequest.filters,
+      getConfig,
+      sort: body.sort,
+    });
 
     const bodyToReturn = {
       ...searchRequest.body,
       pit: searchRequest.pit,
+      query: builtQuery,
+      highlight:
+        searchRequest.highlightAll && builtQuery
+          ? getHighlightRequest(getConfig(UI_SETTINGS.DOC_HIGHLIGHT))
+          : undefined,
+      // remove _source, since everything's coming from fields API, scripted, or stored fields
+      _source: fieldListProvided && !sourceFieldsProvided ? false : body._source,
+      stored_fields:
+        fieldListProvided || sourceFieldsProvided ? [...new Set(remainingFields)] : ['*'],
+      runtime_mappings: runtimeFields,
+      script_fields: scriptedFields,
+      fields: this.getFieldsList({
+        index,
+        fields,
+        docvalueFields: body.docvalue_fields,
+        fieldsFromSource,
+        filteredDocvalueFields,
+        metaFields,
+        fieldListProvided,
+        sourceFieldsProvided,
+      }),
     };
 
-    return omitByIsNil({ ...searchRequest, body: omitByIsNil(bodyToReturn) }) as SearchRequest;
+    return omitByIsNil({
+      ...searchRequest,
+      body: omitByIsNil(bodyToReturn),
+      indexType: this.getIndexType(index),
+      highlightAll:
+        searchRequest.highlightAll && builtQuery ? undefined : searchRequest.highlightAll,
+    }) as SearchRequest;
+  }
+
+  private getFieldFilter({
+    bodySourceExcludes,
+    metaFields,
+  }: {
+    bodySourceExcludes: any;
+    metaFields: any;
+  }) {
+    const filter = fieldWildcardFilter(bodySourceExcludes, metaFields);
+    return (fieldsToFilter: any) =>
+      fieldsToFilter.filter((fld: SearchFieldValue) => filter(this.getFieldName(fld)));
+  }
+
+  private getUniqueFieldNames({
+    fields,
+    fieldsFromSource,
+  }: {
+    fields: any;
+    fieldsFromSource: any;
+  }) {
+    const bodyFieldNames = fields.map((field: SearchFieldValue) => this.getFieldName(field));
+    return [...new Set([...bodyFieldNames, ...fieldsFromSource])];
+  }
+
+  private filterScriptFields({
+    uniqFieldNames,
+    scriptFields,
+  }: {
+    uniqFieldNames: any;
+    scriptFields: any;
+  }) {
+    return uniqFieldNames.includes('*')
+      ? scriptFields
+      : // filter down script_fields to only include items specified
+        pick(
+          scriptFields,
+          Object.keys(scriptFields).filter((f) => uniqFieldNames.includes(f))
+        );
+  }
+
+  private getBuiltEsQuery({
+    index,
+    query,
+    filters,
+    getConfig,
+    sort,
+  }: {
+    index: any;
+    query: any;
+    filters: any;
+    getConfig: (key: string) => any;
+    sort?: any;
```

Comment:
This will specifically be addressed in a follow up PR. This PR makes a number of preexisting issues more explicit but its still a pretty big improvement.

Reply:
I'll look through this PR for easy typescript improvements. There should be at least a couple. Anything that gets too complicated will happen in a follow up.